server/src/main/scala/org/http4s/server/middleware/RequestId.scala

```diff
@@ -38,38 +38,37 @@ object RequestId {
 
   val requestIdAttrKey: Key[String] = Key.newKey[IO, String].unsafeRunSync()
 
-  def apply[G[_], F[_]](http: Http[G, F])(implicit G: Sync[G]): Http[G, F] =
+  def apply[G[_], F[_]](
+      http: G[Response[F]])(implicit G: Sync[G], L: Local[G, Request[F]]): G[Response[F]] =
     apply(requestIdHeader)(http)
 
   def apply[G[_], F[_]](
       headerName: CIString
-  )(http: Http[G, F])(implicit G: Sync[G]): Http[G, F] =
-    Kleisli[G, Request[F], Response[F]] { req =>
-      for {
-        header <- req.headers.get(headerName).map(_.head) match {
-          case None => G.delay(Header.Raw(headerName, UUID.randomUUID().toString()))
-          case Some(header) => G.pure(header)
-        }
-        reqId = header.value
-        response <- http(req.withAttribute(requestIdAttrKey, reqId).putHeaders(header))
-      } yield response.withAttribute(requestIdAttrKey, reqId).putHeaders(header)
-    }
+  )(http: G[Response[F]])(implicit G: Sync[G], L: Local[G, Request[F]]): G[Response[F]] =
+    for {
+      req <- L.ask
+      header <- req.headers.get(headerName).map(_.head) match {
+        case None => G.delay(Header.Raw(headerName, UUID.randomUUID().toString()))
+        case Some(header) => G.pure(header)
+      }
+      reqId = header.value
+      resp <- L.local(http)(_.withAttribute(requestIdAttrKey, reqId).putHeaders(header))
```

Comment:
Effectively asking the request a second time, because we effectfully transformed the request.  I haven't counted the flatMaps, but this feels flatMappier.